### PR TITLE
Corrections in the IEEG validation for the MEF3 file format 

### LINKED ISSUE #953: Validator error with iEEG data in MEF3 format
I am preparing a BIDS-iEEG dataset to share on OpenNeuro. Some of my iEEG data are in MEF3 format, which should currently be allowed (but not supported) and maybe give a warning, but not an error. The validator does however not recognize the _ieeg.mefd file as BIDS-compatible and gives errors. 

@sappelhoff @maxvandenboom

### PR DESCRIPTION
**Pull request by MaxvandenBoom**
@dorahermes asked me to look into this.
This PR has been tested and closes https://github.com/bids-standard/bids-validator/issues/953

### CODE REVIEW

**Review by sappelhoff — COMMENTED** [review]
Thanks @MaxvandenBoom 

Could you fix the test, please? As a guidance, see how the `.ds` directories of the CTF system in MEG are tested:

https://github.com/bids-standard/bids-validator/blob/0d21b0adc30ab00ba53df61cbbb4d931cbac36ee/bids-validator/tests/type.spec.js#L182-L183

I guess you could more or less copy this style for `.mefd` directories.


This brings me to another point: you added several file extensions to the list of permissible extensions (e.g., rdat, ridx, ...) -> I think that this should not be necessary because to my knowledge the validator does not descend into the `.mefd` directory --> again: see the CTF `.ds` test: the files within `.ds` can be essentially garbage and it's fine to the validator.

**Comment by MaxvandenBoom** [comment]
@sappelhoff I updated the test as you requested.

Before adding the file extensions, I check the `ephysTest` function and the files inside of the `.mefd` pass through there. Since they otherwise end up being added only to `misc`, I figured it made sense have them in 'ephys' as well, they are electrophysiology data after all.
Let me know what you want me to do with this...

File: bids-validator/tests/type.spec.js [hunk 1]
```diff
@@ -292,7 +292,11 @@ describe('utils.type.file.isIEEG', function() {
     '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.set',
     '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.fdt',
     '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.nwb',
-    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd',
+    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/sub-01_ses-001_task-rest_run-01_ieeg.rdat',
+    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/sub-01_ses-001_task-rest_run-01_ieeg.ridx',
+    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.tdat',
+	'/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.idx',
+	'/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.tmet',
```
Comment (sappelhoff) [inline bids-validator/tests/type.spec.js]:
```suggestion
    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.tdat',
    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.idx',
    '/sub-01/ses-001/ieeg/sub-01_ses-001_task-rest_run-01_part-01_ieeg.mefd/CH1.timd/CH1-000000.segd/sub-01_ses-001_task-rest_run-01_ieeg.tmet',
```

**Review by sappelhoff — COMMENTED** [review]
>  Since they otherwise end up being added only to misc, I figured it made sense have them in 'ephys' as well

you are right, thanks for checking. I was initially worried that now any file with the extensions like "rdat", "ridx", etc. could be added even without being part of a  `.mefd` directory, but that's not true :slightly_smiling_face: 

I left one suggestion to fix a linter error.

Otherwise there is only the code climate error that now apparently some code is "too complex". If you know how to fix this, feel free to do it -- but from my side we can merge this as soon as the linter error is gone.

**Comment by MaxvandenBoom** [comment]
Ah my mistake, I work cross OSs, it sometimes becomes messy with tabs and whitespaces characters.
I accepted you commit

**Comment by MaxvandenBoom** [comment]
@sappelhoff I could definitely fix the lintel error, they all come from the `verifyDatafileMatch` function  I was considering various optimizations to that function anyway, but felt it was not my place. Let me see what I can do

**Comment by MaxvandenBoom** [comment]
@sappelhoff I tried 3 ways of writing it (see past commits), but codeclimate is being unreasonable. I don't know how to satisfy it's "Cognitive Complexity" measure without re-writing the function in a very long, messy and inefficient way.

Can you maybe help?

**Comment by MaxvandenBoom** [comment]
PS. The current state (last commit) is a pretty reasonable and working solution to accept and merge...

**Review by sappelhoff — APPROVED** [review]
I agree the PR seems very reasonable to me now. Thanks a lot!

@rwblair please merge if you are happy.

**Comment by sappelhoff** [comment]
re: codeclimate --> thanks for trying. I remember fighting this thing in some PRs I had in the past ... and it's really tricky.

File: bids-validator/utils/bids_files.js [hunk 2]
```diff
@@ -56,24 +56,31 @@ function checkFileListForMatch(i) {
...
 
-  // MEG datafiles may be a folder, therefore not contained in fileList, will need to look in paths
+  // IEEG and MEG datafiles may be a folder, therefore not contained in fileList, will need to look in paths
   if (
     !isSelf &&
-    !match &&
-    (noExt.endsWith('_meg') || noExt.endsWith('_coordsystem'))
-  ) {
-    megDs = matchFile.relativePath.includes('_meg.ds')
+    !match) {
```
Comment (effigies) [inline bids-validator/utils/bids_files.js]: This no longer needs to be multi-line. Can we change to:

```Javascript
if (!isSelf && !match) {
```

File: bids-validator/utils/bids_files.js [hunk 3]
```diff
@@ -56,24 +56,31 @@ function checkFileListForMatch(i) {
...
+
+    if (noExt.endsWith('_meg') || noExt.endsWith('_coordsystem')) {
+      folderMatch = matchFile.relativePath.includes('_meg.ds')
+    }
+    if (noExt.endsWith('_ieeg') || noExt.endsWith('_coordsystem')) {
+      folderMatch = matchFile.relativePath.includes('_ieeg.mefd')
+    }
+
```
Comment (effigies) [inline bids-validator/utils/bids_files.js]: I think ESLint is [complaining](https://app.circleci.com/pipelines/github/bids-standard/bids-validator/92/workflows/302a4a6f-c322-4405-9175-8904960564bf/jobs/4961/steps) about these extra newlines:

```suggestion
    if (noExt.endsWith('_meg') || noExt.endsWith('_coordsystem')) {
      folderMatch = matchFile.relativePath.includes('_meg.ds')
    }
    if (noExt.endsWith('_ieeg') || noExt.endsWith('_coordsystem')) {
      folderMatch = matchFile.relativePath.includes('_ieeg.mefd')
    }
```

**Comment by MaxvandenBoom** [comment]
@effigies Thanks for the suggestions. I applied them both, but codeclimate is still equally confused :)

File: bids-validator/utils/bids_files.js [hunk 4]
```diff
@@ -56,24 +56,27 @@ function checkFileListForMatch(i) {
...
 
-  // MEG datafiles may be a folder, therefore not contained in fileList, will need to look in paths
-  if (
-    !isSelf &&
-    !match &&
-    (noExt.endsWith('_meg') || noExt.endsWith('_coordsystem'))
-  ) {
-    megDs = matchFile.relativePath.includes('_meg.ds')
+  // IEEG and MEG datafiles may be a folder, therefore not contained in fileList, will need to look in paths
+  if (!isSelf &&!match) {
```
Comment (sappelhoff) [inline bids-validator/utils/bids_files.js]:
```suggestion
  if (!isSelf && !match) {
```

this would fix at least the linter

**Comment by sappelhoff** [comment]
code climate also makes a bet on how long it'll probably take to fix these issues. I find that hilarious:

image

**Comment by MaxvandenBoom** [comment]
> code climate also makes a bet on how long it'll probably take to fix these issues. I find that hilarious:
> 
> image

Haha, yes I saw those. I guess codeclimate will keep you busy fixing the issue until that estimate becomes true :)

**Comment by sappelhoff** [comment]
I am merging this now, because codeclimate is an advisor, not a dictator :-)

thanks a lot @MaxvandenBoom!

**Pull request merged**